File: ruby/allergies/allergies.rb
class Allergies
  attr_reader :list

  def initialize(coded_allergens)
    @list = ALLERGENS.select.with_index do |_, left_shift|
      coded_allergens.anybits?(1 << left_shift)
    end
  end

  def allergic_to?(allergen)
    list.include?(allergen)
  end

  private
  ALLERGENS = %w[eggs peanuts shellfish strawberries tomatoes chocolate pollen cats].freeze
  private_constant :ALLERGENS
end

File: elixir/allergies/lib/allergies.ex
defmodule Allergies do
  use Bitwise

  @allergens ~w[eggs peanuts shellfish strawberries tomatoes chocolate pollen cats]

  def list(flags) do
    Enum.filter(@allergens, &allergic_to?(flags, &1))
  end

  def allergic_to?(flags, item) do
    (flags &&& 1 <<< Enum.find_index(@allergens, &(&1 == item))) !== 0
  end
end
